File: example/flux_example_executor/src/split_sub.cpp
#include "flux/ros/executor.hpp"
#include "flux/ros/subscription.hpp"

#include <rclcpp/rclcpp.hpp>

#include "sensor_msgs/flux/image.hpp"
#include <sensor_msgs/msg/image.hpp>

#include <chrono>
#include <cstdint>
#include <memory>
#include <mutex>
#include <thread>

// Two executors, two threads: a stock rclcpp one for the ROS side and a flux one for the flux
// side. This is the arrangement merged_sub does NOT use, and it is available because a flux
// subscription is not an rclcpp entity -- it is in no callback group, so nothing here contends
// for a group claim.
//
// What it costs is the mutual exclusion merged_sub gets for free. The two callbacks run on
// different threads at the same time, so anything both touch needs a lock. That lock is the
// whole reason to look at this file before choosing this shape.
class SplitSubscriber : public rclcpp::Node
{
public:
  SplitSubscriber()
  : Node("flux_split_sub"),
    flux_sub_(*this, kTopic, Image::kFingerprint, [this](const flux::FrameView & f) { on_flux(f); })
  {
    ros_sub_ = create_subscription<sensor_msgs::msg::Image>(
      kTopic, rclcpp::QoS(rclcpp::KeepLast(kDepth)).best_effort(),
      [this](const sensor_msgs::msg::Image::ConstSharedPtr & msg) { on_ros(*msg); });
    timer_ = create_wall_timer(std::chrono::seconds(1), [this]() { report(); });
  }

  flux::ros::Subscription & flux_subscription() { return flux_sub_; }

private:
  using Image = sensor_msgs::flux_msg::Image;

  static constexpr const char * kTopic = "image";
  static constexpr std::size_t kDepth = 8;

  void on_flux(const flux::FrameView & f)
  {
    Image::View v{f};
    const std::uint32_t width = v.width();
    if (!v.ok__()) {
      return;
    }
    const std::lock_guard<std::mutex> lock(mutex_);  // the ROS thread reads these too
    flux_width_ = width;
    ++flux_seen_;
  }

  void on_ros(const sensor_msgs::msg::Image & msg)
  {
    const std::lock_guard<std::mutex> lock(mutex_);
    ros_width_ = msg.width;
    ++ros_seen_;
  }

  void report()
  {
    const std::lock_guard<std::mutex> lock(mutex_);
    RCLCPP_INFO(
      get_logger(), "flux %lu (w=%u)  ros %lu (w=%u)", flux_seen_, flux_width_, ros_seen_,
      ros_width_);
  }

  flux::ros::Subscription flux_sub_;
  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr ros_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
  std::mutex mutex_;
  std::uint64_t flux_seen_ = 0;
  std::uint64_t ros_seen_ = 0;
  std::uint32_t flux_width_ = 0;
  std::uint32_t ros_width_ = 0;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<SplitSubscriber>();

  // The node goes to the rclcpp executor. The flux executor is given the flux subscription and
  // nothing else -- handing the node to both is what would throw.
  flux::ros::Executor fex;
  fex.add(node->flux_subscription());

  rclcpp::on_shutdown([&fex]() { fex.stop(); });

  std::thread ros_thread([node]() { rclcpp::spin(node); });
  fex.spin();
  ros_thread.join();

  rclcpp::shutdown();
  return 0;
}

File: example/flux_example_executor/flux_example_executor/split_sub.py
#!/usr/bin/env python3
"""Give the ROS side and the flux side an executor each, on two threads.

ros2 run flux_example_executor split_sub.py

merged_sub puts both on one thread and gets mutual exclusion for free. Here they run at the same
time, so anything both callbacks touch needs a lock. That lock is the cost of the shape.

This works because a flux subscription is not an rclpy entity: it is in no callback group, so the
two executors never contend for one. The flux executor is given the flux subscription and nothing
else: no add_ros_node, because rclpy.spin already has the node.
"""

import threading

import flux.ros
import rclpy
from rclpy.executors import ExternalShutdownException
from rclpy.node import Node
from rclpy.qos import QoSProfile
from rclpy.qos import ReliabilityPolicy
from sensor_msgs.msg import Image as RosImage
from sensor_msgs_flux.image import Image


class SplitSubscriber(Node):

    TOPIC = "image"
    DEPTH = 8

    def __init__(self):
        super().__init__("flux_split_sub")
        self.lock = threading.Lock()
        self.flux_seen = 0
        self.ros_seen = 0
        self.flux_width = 0
        self.ros_width = 0

        self.flux_sub = flux.ros.Subscription(
            self, self.TOPIC, callback=self.on_flux, fingerprint=Image.FINGERPRINT__
        )
        qos = QoSProfile(depth=self.DEPTH, reliability=ReliabilityPolicy.BEST_EFFORT)
        self.ros_sub = self.create_subscription(RosImage, self.TOPIC, self.on_ros, qos)
        self.timer = self.create_timer(1.0, self.report)

    def on_flux(self, view):
        width = Image.View(view).width
        with self.lock:               # the rclpy thread reads these too
            self.flux_width = width
            self.flux_seen += 1

    def on_ros(self, msg):
        with self.lock:
            self.ros_width = msg.width
            self.ros_seen += 1

    def report(self):
        with self.lock:
            self.get_logger().info(
                f"flux {self.flux_seen} (w={self.flux_width})  "
                f"ros {self.ros_seen} (w={self.ros_width})"
            )


def main(args=None):
    rclpy.init(args=args)
    node = SplitSubscriber()

    # No add_ros_node here: rclpy.spin below has the node, and handing it to both is the mistake
    # this shape exists to avoid.
    flux_executor = flux.ros.Executor()
    flux_executor.add_flux(node.flux_sub)
    flux_thread = threading.Thread(target=flux_executor.spin, daemon=True, name="flux-spin")
    flux_thread.start()
    try:
        rclpy.spin(node)
    except (KeyboardInterrupt, ExternalShutdownException):
        pass
    finally:
        flux_executor.stop()
        flux_thread.join(timeout=2.0)
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()
